runtime/src/vote_account.rs

```diff
@@ -31,6 +49,100 @@ impl VoteAccount {
         });
         self.vote_state.read().unwrap()
     }
+
+    /// VoteState.node_pubkey of this vote-account.
+    fn node_pubkey(&self) -> Option<Pubkey> {
+        Some(self.vote_state().as_ref().ok()?.node_pubkey)
+    }
+}
+
+impl VoteAccounts {
+    pub fn staked_nodes(&self) -> HashMap<Pubkey, u64> {
+        self.staked_nodes_once.call_once(|| {
```

Comment:
Wait, so if `bank.staked_nodes()` is called before the bank is frozen, will `staked_nodes` return an incomplete view of the staked nodes if further transactions update the `bank.vote_accounts`?

Reply:
once `staked_nodes` is calculated, any update to `vote_accounts` also updates the `staked_nodes` cache. plz see `test_staked_nodes` bellow.

Reply:
I see, ok.

Reply:
So if the stake goes to zero, how does it work with the key being present in the staked_nodes? Is that consistent to how it was before?

Reply:
If stake goes to zero, the node is removed from staked nodes.

Previously, if there is a `vote-account` with _zero stake_, then `VoteAccount.vote_state.node_pubkey` would still show up in the staked-nodes hash-map but with zero stake. So this is a change from before.

But a consequence of previous logic is that _sometimes_ (but not always) zero staked nodes are counted as "staked":
https://github.com/solana-labs/solana/blob/9593ba909/core/src/cluster_info.rs#L2085-L2089
https://github.com/solana-labs/solana/blob/9593ba909/core/src/cluster_info.rs#L2049
https://github.com/solana-labs/solana/blob/9593ba909/core/src/crds_gossip_pull.rs#L525

But that seems more like a bug rather than the intended behavior. Other places do not make this distinction for zero-staked nodes whether they show up in the hash map or not. So, unless I am missing something, I think it is the right thing to exclude nodes with zero stake from staked-nodes hash-map.

Do you know of any place that counting zero-staked nodes as staked is actually intentional and desired behavior?

Reply:
Yea, I agree seems like a bug. I think it sounds ok.

Reply:
@behzadnouri thanks for the thorough investigation! I think the same zero-lamport not showing up is true for places where there was a change from `staking_utils::staked_nodes_at_epoch() ->  epoch_staked_nodes()`, so it would be good to double check the zero-staked changes are safe as well.

Reply:
From what I gather:

1) The change in broadcast stage: https://github.com/solana-labs/solana/pull/13929/files#diff-b9b3dfeefcf7a68b5072d1b4fa4beb2598e93ce3130856beab78607cc960c60cR308 should be safe because it ultimately calls down to: https://github.com/solana-labs/solana/blob/master/core/src/cluster_info.rs#L1254-L1256, which accounts for missing stakes

2) The change in retransmit: https://github.com/solana-labs/solana/pull/13929/files#diff-8d73ca7d46b0b2dba59642e0e183fbb6127c8459ade483df4ac729390519e6f6R248, like 1) above, also calls down to https://github.com/solana-labs/solana/blob/master/core/src/cluster_info.rs#L1254-L1256, which accounts for missing stakes

3) The change in leader schedule: https://github.com/solana-labs/solana/pull/13929/files#diff-e7124b3c7b7174e8cc0a9a8a62dc8fbd9e52de923c3133712eaffb96d6c8c414R10 would be scary if it caused the leader schedule to be computed differently, but should be ok if `WeightedIndex` treats zero weight as a no-op: https://github.com/solana-labs/solana/blob/master/ledger/src/leader_schedule.rs#L25, but this would be good to confirm.

Reply:
yes, weighted-index should ignore zero weights